airgap/v2/pkg/database/database.go

```diff
@@ -0,0 +1,105 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package database
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/go-logr/logr"
+	v1 "github.com/redhat-marketplace/redhat-marketplace-operator/airgap/v2/apis/model/v1"
+	"github.com/redhat-marketplace/redhat-marketplace-operator/airgap/v2/pkg/models"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+)
+
+type FileStore interface {
+	SaveFile(finfo *v1.FileInfo, bs []byte) error
+	DownloadFile(finfo *v1.FileID) (*models.Metadata, error)
+}
+
+type Database struct {
+	DB  *gorm.DB
+	Log logr.Logger
+}
+
+// SaveFile allows us to save a file along with it's metadata to the database
+func (d *Database) SaveFile(finfo *v1.FileInfo, bs []byte) error {
+	// Validating input data
+	if finfo == nil || bs == nil {
+		return fmt.Errorf("nil arguments received: finfo: %v bs: %v", finfo, bs)
+	} else if finfo.GetFileId() == nil {
+		return fmt.Errorf("file id struct is nil")
+	} else if len(strings.TrimSpace(finfo.GetFileId().GetId())) == 0 && len(strings.TrimSpace(finfo.GetFileId().GetName())) == 0 {
+		return fmt.Errorf("file id/name is blank")
+	}
+
+	// Create a slice of file metadata models
+	var fms []models.FileMetadata
+	m := finfo.GetMetadata()
+	for k, v := range m {
+		fm := models.FileMetadata{
+			Key:   k,
+			Value: v,
+		}
+		fms = append(fms, fm)
+	}
+
+	// Create metadata along with associations
+	metadata := models.Metadata{
+		ProvidedId:      finfo.GetFileId().GetId(),
+		ProvidedName:    finfo.GetFileId().GetName(),
+		Size:            finfo.GetSize(),
+		Compression:     finfo.GetCompression(),
+		CompressionType: finfo.GetCompressionType(),
+		File: models.File{
+			Content: bs,
+		},
+		FileMetadata: fms,
+	}
+	err := d.DB.Create(&metadata).Error
+	if err != nil {
+		d.Log.Error(err, "Failed to save model")
+		return err
+	}
+
+	d.Log.Info(fmt.Sprintf("File of size: %v saved with id: %v", metadata.Size, metadata.FileID))
+	return nil
+}
+
+// DownloadFile allows us to extract a file and it's metadata from the database provided it exists
+func (d *Database) DownloadFile(finfo *v1.FileID) (*models.Metadata, error) {
+
+	var meta models.Metadata
+
+	fileid := strings.TrimSpace(finfo.GetId())
+	filename := strings.TrimSpace(finfo.GetName())
+
+	// Perform validations and query
+	if len(fileid) != 0 {
+		d.DB.Where("provided_id = ?", fileid).Order("created_at desc").Preload(clause.Associations).First(&meta)
+	} else if len(filename) != 0 {
+		d.DB.Where("provided_name = ?", filename).Order("created_at desc").Preload(clause.Associations).First(&meta)
+	} else {
+		return nil, fmt.Errorf("file id/name is blank")
+	}
+
+	if reflect.DeepEqual(meta, models.Metadata{}) {
+		return nil, fmt.Errorf("no file found for provided_name: %v / provided_id: %v", filename, fileid)
+	}
+	d.Log.Info(fmt.Sprintf("Retreived file of size: %v with id: %v", meta.Size, meta.FileID))
```

Comment:
```suggestion
	d.Log.Info("Retreived file of size", "size", meta.Size, "id", meta.FileID))
```

logr.Logger interface should allow this, it's a little cleaner and can be searched in log output easier.